### PR TITLE
Adding in support for credential structures

### PR DESCRIPTION
**Pull request by stabby-r7**
Initial POC PR for adding in support for IDR-like credential structures into the SDK.

Naturally, this barely helps since it's go, and they're all in python, but we need to start somewhere.

This will also require changes in prime komand.

Ross, you mentioned wanting to pair this with a type:credential and structure: (password | asymmetric_key | token) bundle to determine what cred we use. In doing this, it seemed like those 3 could just be regular types (Yes, i know, the word type is loaded - but if the idea of the structure is only to get them over the word 'type', then I think we don't actually need it).

I could see where we already have a thing called password, but we could simply have our "type" for them be `credential_x` or `credential_y`, and avoid the need for structure. Then we just need a mapping between their type and what we call it, which we can manage elsewhere.

Either way, we carry some additional complexity - either here in the original generation, or later on when we need to make sense of it. Probably here, in the generation, is the right move, but I wanted to ask the question all the same.

### CODE REVIEW

**Comment by stabby-r7** [comment]
Oh, haven't tested yet - doing so now, wanted to just get something up to start the conversation

**Comment by stabby-r7** [comment]
Looking at the equivalent code in Komand...

We may have a bit of an issue with the approach of using indirection to parse out the "real" type from a pair of "credential + structure".

The parser is written in a way to try and mimic a low level language parser, of sorts, where we simply have a scanner that is reading bytes in a straight line. This is not conducive to parsing things neatly in a single pass - I need to look into it a bit to see if we can make use of the "structure" concept at a higher level, and then swap out the "credential" placeholder type with the real one based on the structure.

**Comment by stabby-r7** [comment]
Alright luckily, the code in old Komand around this only looks a certain way - we wrapped simple stuff in a complex wrapper, then wrapped that in another, simple wrapper.

So i think I can plumb my way around the issue.

File: type_mapper.go [hunk 1]
```diff
@@ -32,17 +42,22 @@ func NewTypeMapper(s *PluginSpec) *TypeMapper { // TO FIGURE OUT object and byte
...
+	strippedType := pd.Type
+	if strings.HasPrefix(pd.Type, "[]") {
+		strippedType = pd.Type[2:]
 		isArr = true
 	}
 	foundType, ok := t.CustomTypes[strippedType]
 	if !ok {
-		// We don't have a remap for it, so use the original
-		return specType
+		return pd.Type
```
Comment (mbroomfield-r7) [inline type_mapper.go]: Looks like a comment got swollowed

**Comment by mbroomfield-r7** [comment]
I've been thinking about what would be the best UX for the plugin developer.
Credential types are similar to file types. They're predefined (and also slightly opaque) structures.

Option A)
```
my_param:
  type: credential
  structure: credential.Password
```

Option B)
```
my_param:
  type: credential_basicauth
```

If I'm a plugin dev and I've done file types, then the concept of Option B, just using another predefined structure, is easier to understand out of the gate. Unlike option A, which requires a developer to know about another field `structure`. And they also have to know the value of the `structure` field.

Now one concerned voiced by stabby is somewhere there might have to be a mapping from name to structure for the credential types. In both cases, I can't see how you get around keeping *something* in komand. (What I mean here is running in the cloud, not which dependency the map ends up in.)

So if I had to choose, Option B seems way better from a UX perspective.

Actually you could drop the prefix `credential_` from the type. Just say `basicauth` or `oauth` or whatever.

**Comment by mrinehart-r7** [comment]
So the way I'm interpreting this leads me to believe this is adding a new usable type for plugin developers, specifically for authentication? And it would have (reading the code) different sets of common authentication schemes?

Is this is a step towards unifying connection/input in plugins?

**Comment by mbroomfield-r7** [comment]
Just talked to ryan about this, changing my opinion. Please stand by

**Comment by mbroomfield-r7** [comment]
I talked with ryan about the requirements around entering connections and entering credentials (which are two different UI boxes), and there's actually a few reasons why we might want to go with Option A anyways.

Well actually I'll call it Option C:
```
my_param:
  type: credential
  structure: [basicauth, oauth]
```

If we format plugin specs like this, we solve another current issue with plugins: A connection might take more than 1 type of credential. Currently, our solution to this is to have many unused fields in the connection.

TOBECLEAR: this is not a Q2 requirement so no need to make `structure` a list right now.
Also, all `type` fields could be lists as well. But I think if we ever have lists the `structure` field feels the cleanest.

So to sum it up: I think this code as is is fine, aka Option A is an ok way to go. Sure devs have to learn about a new field, but it makes things easier for the product to understand + it has the potential to be expanded to lists which solve the "many valid cred structures" problem.

**Comment by mbroomfield-r7** [comment]
Ok so I have my final thoughts.

We should go with the `type: credential_basicauth` solution which I was first in favor for.

I think we get all the things we want:
I added a more lengthy comment about this here:
https://github.com/komand/komand/pull/1878#discussion_r190218789

**Comment by stabby-r7** [comment]
So, going to close this and re-open a new branch based on what we've settled on, which is much more streamlined after talking to jen, ross, ryan, and the sweeper.

Ryans concern: can he detect credential types easily.

Sweeps concern: less stuff for developers to mentally remember, prevent people from doing the wrong thing when making their own types.

I think we can accomplish all of these things, especially now that we've gotten CC team to agree to a rename on the password type.
